### DISCUSSION TITLE
Parse streams with several packets and non-json elements

### DISCUSSION THREAD

**Discussion by AlexanderVasiljev** [nlohmann/json#3133] [Q&A]
Hi,

I need to parse continuous streams with json packets. At the moment of parsing a stream can contain several packets, not complete packets and not json stuff.

For example, the stream may come up with two strings
```json
58{"cmd": "version"}{"cmd": "dac"}{"cmd":
"delay"}fg{"cmd": "wait"}
```
It should be parsed into 4 json objects.
Here is the code that I use

```cpp
void parse(std::stringstream& ss, std::vector<json> &jv)
{
    std::size_t last = 0;
    while(!ss.eof())
    {
        try
        {
            json j1;
            ss >> j1;
            jv.push_back(j1);
        }
        catch(json::parse_error& ex)
        {
            std::cerr << ex.what() << std::endl;
	    if(ss.eof())
	   	 last = ex.byte;
  	    else if(ex.byte > 1)
		ss.unget();
        }
    }
    ss.seekg(-last+1, std::ios_base::end);
}

std::string str = "58{\"cmd\": \"version\"}{\"cmd\": \"dac\"}{\"cmd\":";
std::stringstream ss;
std::vector<json> jv;
ss << str;
parse(ss, jv);
ss << "\"delay\"}fg{\"cmd\": \"wait\"}po";
parse(ss, jv);

for(const auto& j : jv)
    std::cout << "json = " << j << std::endl;
```

And the output is

> 
> [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected ':'; expected '[', '{', or a literal
> [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected '}'; expected '[', '{', or a literal
> [json.exception.parse_error.101] parse error at line 1, column 8: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal
> [json.exception.parse_error.101] parse error at line 1, column 2: syntax error while parsing value - invalid literal; last read: 'fg'
> [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'g'
> [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal
> 
> json = 58
> json = "cmd"
> json = "version"
> json = {"cmd":"dac"}
> json = {"cmd":"delay"}
> json = {"cmd":"wait"}
> 

There are 6 objects instead of 4.
Why is the "58" parsed as valid object and why is the next json packet not parsed properly? But the "fg" is ignored as expected.

**Comment by nlohmann** [comment]
This resync approach is fragile because ex.byte (the offset reported in a parse_error) does not map cleanly onto the underlying std::stringstream's read position: the lexer/input adapter reads through an internal buffered iterator with lookahead, so after an exception the stream may already be positioned past where ex.byte suggests, and unget()/seekg() calls based on ex.byte can under- or over-consume characters - which is exactly why '58' gets accepted as a standalone JSON number (a bare integer is valid top-level JSON) and why the recovery skips or duplicates tokens. nlohmann/json's parser isn't designed to auto-resynchronize on garbage embedded between values; it expects each value it's asked to parse to be well-formed JSON (or, for JSON Lines, newline-delimited - see https://json.nlohmann.me/features/parsing/json_lines/, though that page notes operator>> doesn't work for that case either). For a stream that can contain garbage or partial packets between JSON messages, you'll get more reliable results by adding explicit framing to your protocol (e.g. a length prefix or a delimiter guaranteed not to appear inside a JSON value) and only handing complete, delimited chunks to json::parse(), rather than relying on catching parse_error and reinterpreting the byte offset to resynchronize.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
